source/extensions/transport_sockets/http_11_proxy/connect.cc

```diff
@@ -0,0 +1,140 @@
+#include "source/extensions/transport_sockets/http_11_proxy/connect.h"
+
+#include <sstream>
+
+#include "envoy/network/transport_socket.h"
+
+#include "source/common/buffer/buffer_impl.h"
+#include "source/common/common/scalar_to_byte_vector.h"
+#include "source/common/common/utility.h"
+#include "source/common/network/address_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Http11Connect {
+
+const std::string& UpstreamHttp11ConnectSocket::key() {
+  CONSTRUCT_ON_FIRST_USE(std::string, "envoy.network.transport_socket.http_11_proxy.address");
+}
+
+UpstreamHttp11ConnectSocket::UpstreamHttp11ConnectSocket(
+    Network::TransportSocketPtr&& transport_socket,
+    Network::TransportSocketOptionsConstSharedPtr options)
+    : PassthroughSocket(std::move(transport_socket)), options_(options) {
+  if (options_ && options_->proxyInfo() && transport_socket_->ssl()) {
```

Comment:
I don't think I have a clear picture of how this interacts with socket reuse for multiple requests. For example if the first request via a proxy is for an http:// URL and the second is for an https:// URL, the first request should be sent with a fully qualified URL and no CONNECT, but the second should be a CONNECT request (and then a non-fully qualified URL). Is such a thing possible in Envoy?